craft_parts/executor/part_handler.py

```diff
@@ -187,6 +194,26 @@ def _migrate(
             src_path = str(Path(sub_path) / f)
             self.migrated_directories[dst_partition][src_path] = f
 
+    def _filter_already_distributed(
+        self, visible_files: set[str], visible_dirs: set[str]
+    ) -> tuple[set[str], set[str]]:
+        """Exclude files in sub path already distributed to another partition."""
+        files: set[str] = set()
+        dirs: set[str] = set()
+        if not self._sub_paths_distributed:
+            return visible_files, visible_dirs
+        for f in visible_files.copy():
+            for d in self._sub_paths_distributed:
+                if not Path(f).is_relative_to(Path(d)) or Path(f) == Path(d):
```

Comment:
This checks if the visible file was not already distributed (because it's under a distributed subpath, or it's the subpath itself). Maybe adding a small utility function `_not_already_distributed()` could make this easier to read/understand.

Reply:
A comment paragraph explaining what we're trying to achieve here can be useful.